### ISSUE TITLE
PyTorch 2.0 MPI backend not properly initialized

### ISSUE THREAD

**Issue by roywei** [pytorch/pytorch#97507]
### 🐛 Describe the bug

Build PyTorch 2.0 from source with USE_MPI=1 to enable MPI backends

Run the following code

create the following `test_backends.py` file

```
import sys
import os
import torch
import torch.distributed as dist

def main(backend):
    if backend == "mpi":
        os.environ["RANK"] = os.environ["OMPI_COMM_WORLD_RANK"]
        os.environ["LOCAL_RANK"] = os.environ["OMPI_COMM_WORLD_LOCAL_RANK"]
        os.environ["WORLD_SIZE"] = os.environ["OMPI_COMM_WORLD_SIZE"]
    dist.init_process_group(backend, init_method="env://")
    print("Testing {} backend: hello from rank {}, world size {} ".format(backend, dist.get_rank(), dist.get_world_size()))
    dist.destroy_process_group()

if __name__ == '__main__':
    main(sys.argv[1])
```
MPI backend always return -1 and -1 as rank and world size

```
mpirun -np 2 python test_backends.py mpi
Testing mpi backend: hello from rank -1, world size -1
Testing mpi backend: hello from rank -1, world size -1
```

NCCL and GLOO backends works fine
```
torchrun --nproc_per_node 2 --nnodes 1 test_backends.py nccl
WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
*****************************************
Testing nccl backend: hello from rank 1, world size 2
Testing nccl backend: hello from rank 0, world size 2
```

```
torchrun --nproc_per_node 2 --nnodes 1 test_backends.py gloo
WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
*****************************************
Testing gloo backend: hello from rank 1, world size 2
Testing gloo backend: hello from rank 0, world size 2
```

Build 1.13.1 release and enable USE_MPI works fine
```
Testing mpi backend: hello from rank 0, world size 4 
Testing mpi backend: hello from rank 1, world size 4 
Testing mpi backend: hello from rank 2, world size 4 
Testing mpi backend: hello from rank 3, world size 4 
```

### Versions

2.0.0

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by roywei** [comment]
@H-Huang could you help take a look? thanks!

here is a 1.13.1 pip wheel with MPI enabled, and it works fine

```
pip install https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.13.1_ec2/20221219-193736/54406b8eed7fbd61be629cb06229dfb7b6b2954e/torch-1.13.1%2Bcu117-cp39-cp39-linux_x86_64.whl
```
```
mpirun -np 2 python test_backends.py mpi
Testing mpi backend: hello from rank 1, world size 2
Testing mpi backend: hello from rank 0, world size 2
```

**Comment by ddkalamk** [comment]
@roywei I also faced same issue.

A little debugging revealed that problem is with this line
https://github.com/pytorch/pytorch/blob/v2.0.0/torch/distributed/distributed_c10d.py#L972

MPI backend has it's own way to find out `rank` and `size`, so input `rank` and `size` are set to `-1`. However, with new `backend` based implementation `ProcessGroup` is just a wrapper and actual rank and size are set for `backend_class` and not for `ProcessGroup` wrapper.
One workaround is to recreate `ProcessGroup` after `backend_class` is created in line https://github.com/pytorch/pytorch/blob/v2.0.0/torch/distributed/distributed_c10d.py#L986

```python
backend_class = ProcessGroupMPI.create(global_ranks_in_group)
backend_type = ProcessGroup.BackendType.MPI
if not backend_class:
    return GroupMember.NON_GROUP_MEMBER
# FIX: recreate ProcessGroup with correct rank and size
pg = ProcessGroup(prefix_store, backend_class.rank(), backend_class.size(), base_pg_options)
```

**Comment by roywei** [comment]
Thank you @ddkalamk! it's good to have someone confirm the issue. Will try your fix. We are considering other backends so we 
 don't have to modify and rebuild PyTorch. Looking into this [comparison](https://mlbench.github.io/2020/09/08/communication-backend-comparison/#:~:text=Key%20differences,times%20is%20higher%20than%20MPI's.).

**Comment by rohan-varma** [comment]
cc @H-Huang

**Comment by malfet** [comment]
This test supposed to be run in CI, but even though the issue has been closed long time ago, test was never re-enabled
https://github.com/pytorch/pytorch/blob/97711ac6dba11f1d301cfa51e269cc72d2ab9a9f/.ci/pytorch/test.sh#L595-L600

**Comment by atalman** [comment]
@H-Huang could you please take a look at this issue and provide some feedback

**Comment by atalman** [comment]
cc @H-Huang @roywei 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176

**Comment by H-Huang** [comment]
Hi @roywei, thanks for reporting this issue. This is indeed a regression and the fix (https://github.com/pytorch/pytorch/pull/98545) will be added to 2.0.1
